scrapy_po/utils.py

```diff
@@ -1,7 +1,122 @@
 # -*- coding: utf-8 -*-
+import inspect
+from typing import Tuple, Dict, Type, Callable
+
+import andi
+from scrapy.http import Response
+from scrapy.utils.defer import maybeDeferred_coro
+from twisted.internet.defer import inlineCallbacks, returnValue
+
+from scrapy_po.webpage import Injectable
+from scrapy_po.page_input_providers import providers
+
 
 def get_callback(request, spider):
-    """ Get request.callback of a scrapy.Request, as a callable """
+    """Get request.callback of a scrapy.Request, as a callable."""
     if request.callback is None:
         return getattr(spider, 'parse')
     return request.callback
+
+
+class DummyResponse(Response):
+
+    def __init__(self, url, request=None):
+        super(DummyResponse, self).__init__(url=url, request=request)
+
+
+def is_callback_using_response(callback: Callable):
+    """Check whether the request's callback method is going to use response."""
+    spec = inspect.getfullargspec(callback)
+    try:
+        arg_name = spec.args[1]  # first index is self, second is response
+    except IndexError:
+        # Parse method is probably using *args and **kwargs annotation.
+        # Let's assume response is going to be used.
+        return True
+
+    if arg_name not in spec.annotations:
+        # There's no type annotation, so we're probably using response here.
+        return True
+
+    if issubclass(spec.annotations[arg_name], DummyResponse):
+        # Type annotation is DummyResponse, so we're probably NOT using it.
+        return False
+
+    # Type annotation is not DummyResponse, so we're probably using it.
+    return True
+
+
+def are_dependencies_using_response(plan: andi.Plan):
+    """Check whether any injectable provider makes use of a valid Response."""
+    for obj, _ in plan:
+        provider = providers.get(obj)
+        if not provider:
+            # Provider not found.
+            continue
+
+        spec = inspect.getfullargspec(provider)
```

Comment:
I haven't checked this in detail, but it can be better to use andi.inspect here, because this way you can work around some attr.s-specific issues, and potentially support Union

Reply:
Output is quite different :/

```(Pdb) inspect.getfullargspec(provider)
FullArgSpec(args=['self', 'response'], varargs=None, varkw=None, defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={'response': <class 'scrapy.http.response.Response'>})
(Pdb) andi.inspect(provider)
{'response': []}
```

Reply:
Unlike andi.plan, you need to pass `__init__` method explicitly:

```
In [1]: import andi

In [2]: from scrapy.http  import Response

In [3]: class Foo:
   ...:     def __init__(self, resp: Response):
   ...:         pass
   ...:

In [4]: andi.inspect(Foo)
Out[4]: {'resp': []}

In [5]: andi.inspect(Foo.__init__)
Out[5]: {'resp': [scrapy.http.response.Response]}
```

Reply:
I actually wonder if this logic (https://github.com/scrapinghub/andi/blob/e26bcfd9f946aead542cacac1fa4462af7d01443/andi/andi.py#L289) should be moved to andi.inspect, so that there is less surprises. What do you think @ivanprado ?

Reply:
@kmike I've just refactored this method to make use of `andi.inspect` instead of `inspect.getfullargspec` as you had requested.